### ISSUE TITLE
Implement a way to display one-off notifications in the main dashboard 

### ISSUE THREAD

**Issue by marrrmarrr** [google/site-kit-wp#1110]
## Feature Description

Let's add a way to communicate with Site Kit users about important new features, known issues, etc. exactly where they need to know this information -- in the main dashboard. 

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria
* Build a mechanism to display remote-controlled notifications in the main dashboard 
* The notifications should appear at the top of the dashboard 
* If the user clicks the CTA on the notice → store that information and don’t show it again
* If the user dismisses the notice → store that information and don’t show it again
* If the user does not interact with the notice → display the notice for maximum 3 impressions
* Track clicks on “Participate” button and clicks on “dismiss” 

## Implementation Brief

**PHP:**
* Introduce class `Core\Notifications\Notification` (for inspiration look at `Core\Admin\Notice`):
    * Its constructor should expect `$slug` and `array $args`. The default `$args` should be:
        * `title`: empty string
        * `content`: empty string
        * `image`: empty string
        * `cta_url`: empty string
        * `cta_label`: empty string
        * `cta_target`: "_self"
        * `learn_more_url`: empty string
        * `learn_more_label`: empty string
        * `dismissible`: boolean `false`
        * `dismiss_label: `__( 'Dismiss', 'google-site-kit' )`
    * It should have a `get_slug()` method.
    * It should have a `to_json()` method that returns the full information in an associative array with camel-case keys (e.g. `learn_more_url` would be `learnMoreURL` and `cta_url` would be `ctaURL`), so that this object allows us to have this notification "spec" centrally defined.
* Introduce class `Core\Notifications\Notifications`:
    * Its `register()` method should filter `googlesitekit_rest_routes` to add the following routes:
        * `core/site/data/notifications` should return a list of notification objects queried from `https://sitekit.withgoogle.com/notifications/`. It should instantiate `Notification` objects for every response object.
        * `core/site/data/mark-notification` should expect a `notificationID` and `notificationState` parameter, and then issue a `POST` request to `https://sitekit.withgoogle.com/notifications/mark/` to accept or dismiss a specific notification.
        * The exact specifications of the remote routes can be found separately in the doc.

As a follow-up to this (not per this issue), we should migrate the module notifications endpoint definitions to the `Core\Notifications\Notifications` class as well (as `modules/<module>/data/notifications`), and the API response in `Modules\AdSense` (the only place where we currently have something like this) should be adjusted to use the same paramters via `Core\Notifications\Notification` object.

**JS:**
* Modify `Notification` component in `js/components/notifications` so that it can receive optional (async) `onCTAClick` and `onDismiss` props. If provided, clicking the CTA/dismiss button respectively should first run that callback and afterwards run the existing logic (if any).
* Introduce a `DashboardCoreSiteAlerts` component in `js/components/notifications`, similar to the existing ones. It should fetch the notifications from `core/site/data/notifications` and if one is there return it. It should make use of the `onCTAClick` prop to ensure that when the CTA button is clicked a request to `core/site/data/mark-notification` is made with the ID/slug of that notification (`notificationID`) and the value "accepted" for the `notificationState`. Similarly it should provide `onDismiss` to hit the same route with `notificationState` of "dismissed".
* Modify `js/components/notifications/index.js` with the following filter: `addFilter( 'googlesitekit.DashboardNotifications', 'googlesitekit.SetupNotification', addCoreSiteNotifications, 10 );`.

Given the timeliness of this one, I don't think we have enough time to work on a proper store for these notifications, so I think that should be a follow-up (especially since the `core/site` store is still being worked on right now).

## Changelog entry

* Introduce basic notifications system for information displayed in the Site Kit dashboard.

**Comment by aaemnnosttv** [comment]
How often is `core/site/data/notifications` requested? I assume this is a request we'll want to cache in the browser for some period of time. What about server-side caching? If notifications aren't sensitive or user-specific then we could avoid making too many requests to the API with a short-lived transient.

**Comment by aaemnnosttv** [comment]
Also, where and how will the number of times a notification is shown be tracked? How do we define an impression? (Visible in the viewport for X seconds?)

**Comment by tofumatt** [comment]
JS-wise this seems good to me, especially given the time constraints on this one. I'll leave the PHP to @aaemnnosttv, seems he already has some comments on it 😄

> Given the timeliness of this one, I don't think we have enough time to work on a proper store for these notifications, so I think that should be a follow-up (especially since the core/site store is still being worked on right now).

Agreed on that one for sure.

So IB ✅from me, but Evan's point about request frequency is a good one. I'm not sure how timely we want these notifications to be. It might be that as we move to the data store we do fewer/no PHP proxying of requests, so maybe then we don't have to worry about rate-limiting/caching as much as we do with WordPress.

**Comment by felixarntz** [comment]
@aaemnnosttv 
> How often is `core/site/data/notifications` requested? I assume this is a request we'll want to cache in the browser for some period of time. What about server-side caching? If notifications aren't sensitive or user-specific then we could avoid making too many requests to the API with a short-lived transient.

I think we should request it at the same regularity as regular API data in JS, so once per hour, caching results in session storage. I don't think we need to add PHP caching, as this should work the same way as Google API requests, which we don't cache on the server-side.

> Also, where and how will the number of times a notification is shown be tracked? How do we define an impression? (Visible in the viewport for X seconds?)

Hmm that's a good question. It is tricky for both conceptual and technical reasons:
1. We haven't fully defined when to hide the notification because it has been "ignored". Should it be after x times, once the user logs out (as @marrrmarrr suggested), or never? I'm still leaning towards the latter, as all this notification would ever "clutter" is our own screen - it can easily be dismissed, and even for people who aren't dismissing it, shouldn't get too annoyed because it's not all across WP admin. So I don't see a good reason to hide the notification implicitly without any user interaction.
2. If we decide to hide it after a while without user interaction, the technical question is how to record that the notification was seen. Our current notification system is strange in a way that it only shows one notification at a time. So the API request having been made itself does not indicate the notification being seen, hence server-side tracking is ruled out. And even on the client-side, the component making the request wouldn't really be able to track whether the notification ended up showing due to the current implementation.

**Comment by aaemnnosttv** [comment]
> I don't think we need to add PHP caching, as this should work the same way as Google API requests, which we don't cache on the server-side.

That sounds good. My concern is more about the proxy handling the load than the local WP API. If you're comfortable with that, then so am I 😄 

> We haven't fully defined when to hide the notification because it has been "ignored". Should it be after x times, once the user logs out (as @marrrmarrr suggested), or never? I'm still leaning towards the latter, as all this notification would ever "clutter" is our own screen - it can easily be dismissed, and even for people who aren't dismissing it, shouldn't get too annoyed because it's not all across WP admin. So I don't see a good reason to hide the notification implicitly without any user interaction.

That sounds good to me as well. Building in impression-handling would increase the effort needed here quite a bit, and we can always add that on later if needed. If we were to add it, it might be easier to implement on the proxy API side with a simple time-based expiration rather than building something in on the plugin/client.

Apart from these comments and assuming the impression-based AC is no longer required then the IB looks good to me ✅

**Comment by felixarntz** [comment]
@aaemnnosttv Good to be worked on now, without implicit dismissal support.

**Comment by cole10up** [comment]
**Tested**

Ran a console command (below) to pop the message:

`sessionStorage.setItem('googlesitekit_core::site::notifications', JSON.stringify({value: [{"id":"uxr_study","title":"","content":"We want to hear more about what you think about Site Kit.<br>Sign up to participate in our research studies by telling us a bit more about yourself in the registration form.","image":"","ctaURL":"https://docs.google.com/forms/d/e/1FAIpQLSeh3dCWmg18qHv2f5eidKunVdiSQEIUNeF8EeNYNeZv5XnaLQ/viewform","ctaLabel":"Sure, sign me up","ctaTarget":"_blank","learnMoreURL":"","learnMoreLabel":"","dismissible":true,"dismissLabel":"No thanks"}], date: Date.now() / 1000})) `

**Confirmed palcement below is correct via @aaemnnosttv **
![image](https://user-images.githubusercontent.com/57115907/75521764-3f112280-59bd-11ea-9521-a187ef7e0419.png)

Confirmed browser squeeze test looks good and copy wraps

![image](https://user-images.githubusercontent.com/57115907/75521903-89929f00-59bd-11ea-8bc5-ab59de91a96d.png)

Confirmed selecting buttons stores in the browser cache. 

Confirmed selecting no thanks message disappears. 

Confirmed selecting yes message disappears.

Was unable to fire another test to try 3 interactions as I manually fudged this to display.
